core/templates/rid_owner.h

```diff
@@ -288,11 +288,25 @@ class RID_Alloc : public RID_AllocBase {
 
 	_FORCE_INLINE_ T *get_ptr_by_index(uint32_t p_index) {
 		ERR_FAIL_UNSIGNED_INDEX_V(p_index, alloc_count, nullptr);
+
 		if (THREAD_SAFE) {
 			spin_lock.lock();
 		}
-		uint64_t idx = free_list_chunks[p_index / elements_in_chunk][p_index % elements_in_chunk];
-		T *ptr = &chunks[idx / elements_in_chunk][idx % elements_in_chunk];
+
+		uint32_t index_chunk = p_index / elements_in_chunk;
+		uint32_t index_element = p_index % elements_in_chunk;
+
+		T *ptr = nullptr;
+		// if this index has not been made invalid (freed)
+		if (validator_chunks[index_chunk][index_element] != 0xFFFFFFFF) {
+			uint64_t idx = free_list_chunks[index_chunk][index_element];
+
+			uint32_t idx_chunk = idx / elements_in_chunk;
+			uint32_t idx_element = idx % elements_in_chunk;
+
+			ptr = &chunks[idx_chunk][idx_element];
+		}
+
```

Comment:
That's great! From what I understand, the fix needs to be reworked a little bit though.

Unless I'm missing something, returning nullptr when the element is invalid at the given index would cause to skip an element, so instead it should probably return the next valid element somehow (see get_owned_list for the proper way to access all elements).

Reply:
Ah consequences of writing a PR at 3am .. you are absolutely right, the `alloc_count` returned by `scenario_owner.get_rid_count()` is the number of valid allocations, so we could either do this, or presumably have a function to return `max_alloc`.

I'm actually suspecting that the latter approach is better in some ways, as it makes it easier to iterate through the list and repeat a call `get_ptr_by_index` relatively cheaply. If you wanted to call only the _valid_ ptrs by index, it would presumably be very slow or require a secondary structure to do this.

Quite often I've noticed people using the pattern `get_ptr_by_index(10)` then a line later `get_ptr_by_index(10)`, so it would be nice to keep this cheap (ish).

The only downside is that in the thread safe version, even reading the invalid member contains a spinlock, so in a pathological situation could slow it down. What about having a separate function to return the next valid member of the list? Or we could use a more multi-function external iterator similarish to STL, however this might be overkill.

Looking at `get_owned_list` I remember you mentioning this last night, and it does exactly the same pattern with max_alloc, but getting the whole list each time is not ideal so I think one of these earlier two approaches are most preferable so far...
i.e.
* `next_index = get_next_valid_index(int p_index)`
* or a custom iterator